src/runtime_src/core/common/pskernel_parse.cpp

```diff
@@ -0,0 +1,230 @@
+	    arg.index = index;
+	    arg.type = xrt_core::pskernel::kernel_argument::argtype::global;
+	    ffi_type *enc_ffi_type = &ffi_type_pointer;
+	    arg.ffitype = *enc_ffi_type;
+	    index++;
+	  }
+	  if(dwarf_aggregate_size(type, &var_size) < 0)
+	    throw std::runtime_error("ERROR: Variable Size incorrect!\n");
+
+	  return_args.emplace_back(arg);
+	  break;
+	} while (dwarf_siblingof (&child, &child) == 0);
+
+  return return_args;
+}
+
+// Function to parse object from file on disk
+std::vector<xrt_core::pskernel::kernel_argument> pskernel_parse(const char *so_file, const char *func_name) {
+
+  int fd = open(so_file, O_RDONLY);
+  Dwarf *dw = dwarf_begin(fd, DWARF_C_READ);
+  std::vector<xrt_core::pskernel::kernel_argument> args;
+
+  if (dw != NULL) {
+    Dwarf_Off offset;
+    Dwarf_Off old_offset = 0;
+    size_t hsize;
+    Dwarf_Off abbrev;
+    uint8_t addresssize;
+    uint8_t offsetsize;
+    
+    while (dwarf_nextcu (dw, old_offset, &offset, &hsize, &abbrev, &addresssize, &offsetsize) == 0) {
+      Dwarf_Die cudie_mem;
+      Dwarf_Die *cudie = dwarf_offdie (dw, old_offset + hsize, &cudie_mem);
+      Dwarf_Die child;
+      Dwarf_Die *func_die;
+      
+      if (dwarf_child (cudie, &child) == 0)
+	do {
+	  if(dwarf_tag(&child) == DW_TAG_subprogram) {
+	    const char *name = dwarf_diename (&child);
+	    if (fnmatch(func_name,name,0) == 0) {
+	      func_die = &child;
+	      args = extract_args (func_die);
+	    }
+	  }
+	} while (dwarf_siblingof (&child, &child) == 0);
+      
+      old_offset = offset;
+    }
+    
+    dwarf_end(dw);
+  }
+
+  if(!args.empty())
+    return args;
+  else {
+    throw std::runtime_error("No PS kernel arguments found!");
+    return {};
+  }
+}
+
+// Function to parse from object in memory
+std::vector<xrt_core::pskernel::kernel_argument> pskernel_parse(char *so_file, size_t size, const char *func_name) {
+
+  Elf *ehandle = elf_memory(const_cast<char *>(so_file), size);
+  Dwarf *dw = dwarf_begin_elf(ehandle, DWARF_C_READ, nullptr);
+  std::vector<xrt_core::pskernel::kernel_argument> args;
+  
+  if (dw != NULL) {
+    Dwarf_Off offset;
+    Dwarf_Off old_offset = 0;
+    size_t hsize;
+    Dwarf_Off abbrev;
+    uint8_t addresssize;
+    uint8_t offsetsize;
+    
+    while (dwarf_nextcu (dw, old_offset, &offset, &hsize, &abbrev, &addresssize, &offsetsize) == 0) {
+	Dwarf_Die cudie_mem;
+	Dwarf_Die *cudie = dwarf_offdie (dw, old_offset + hsize, &cudie_mem);
+	Dwarf_Die child;
+	Dwarf_Die *func_die = NULL;
+	
+	if (dwarf_child (cudie, &child) == 0)
+	  do {
+	    if(dwarf_tag(&child) == DW_TAG_subprogram) {
+	      const char *name = dwarf_diename (&child);
+	      if( fnmatch(func_name,name,0)==0 ) {
+		func_die = &child;
+		args = extract_args (func_die);
+	      }
+	    }
+	  } while (dwarf_siblingof (&child, &child) == 0);
+	
+	old_offset = offset;
+      }
+    
+    dwarf_end(dw);
+  }
+
+  if(!args.empty())
```

Comment:
This code should be refactor to something like this:
```
 if(args.empty())
    throw std::runtime_error("No PS kernel arguments found!");

 return args;
```

It just makes it easier to read and straight forward regarding your intent.